### PR TITLE
Feat abbrevs callback

### LINKED ISSUE #3642: abbrevs should only be expanded at the start of the command
## xonfig

<details>

```
+------------------+---------------------+
| xonsh            | 0.9.18              |
| Git SHA          | 26475146            |
| Commit Date      | May 5 06:42:28 2020 |
| Python           | 3.8.3               |
| PLY              | 3.11                |
| have readline    | True                |
| prompt toolkit   | 3.0.5               |
| shell type       | prompt_toolkit      |
| pygments         | 2.6.1               |
| on posix         | True                |
| on linux         | True                |
| distro           | manjaro             |
| on darwin        | False               |
| on windows       | False               |
| on cygwin        | False               |
| on msys2         | False               |
| is superuser     | False               |
| default encoding | utf-8               |
| xonsh encoding   | utf-8               |
| encoding errors  | surrogateescape     |
+------------------+---------------------+
```

</details>

as of abbrevs expands in all places. So it may clash will subcommands. For example, if we have 

~~`aliases['ps'] = 'procs'`~~
```
abbrevs['ps'] = 'procs'
``` 

and do type `docker ps` then it will get expanded to `docker procs`. 

So the default behaviour should be to expand only at the start of the command. special handling for sudo can be added. 
Also we could have special symbols in aliases to mark it as expandable in all places like `aliases['*ps*']` like the glob pattern.

### PR DESCRIPTION
**Pull request by jnoortheen**
fixes #3642

### CODE REVIEW

**Comment by jnoortheen** [comment]
@gforsyth while you are at it, these are good to go as well #4031 #4032

**Comment by gforsyth** [comment]
@laloch -- do you have time to look this over?  I haven't spent much time in the `abbrevs` code

**Comment by jnoortheen** [comment]
@gforsyth this doesn't change any of the existing functionality. Just adds a new feature. Please check the docstring of the module for that added functionality.

**Review by gforsyth — CHANGES_REQUESTED** [review]
Sorry it took me a bit to get to this @jnoortheen -- overall looks good, just two small questions

File: xonsh/xontribs_meta.py [hunk 1]
```diff
@@ -43,11 +46,23 @@ class Xontrib(tp.NamedTuple):
...
     url: str = ""
-    description: str = ""
+    description: tp.Union[str, LazyObject] = ""
     package: tp.Optional[_XontribPkg] = None
     tags: tp.Tuple[str, ...] = ()
 
 
+def get_module_docstring(module: str) -> str:
+    """Find the module and return its docstring without actual import"""
+    print("run get module")
```
[the remark this answers is not in the document]
Comment (jnoortheen) [reply xonsh/xontribs_meta.py:56]: Yes it is. I will remove.

File: xonsh/xontribs_meta.py [hunk 2]
```diff
@@ -43,11 +46,23 @@ class Xontrib(tp.NamedTuple):
...
+    import ast
+    from pathlib import Path
```
[the remark this answers is not in the document]
Comment (jnoortheen) [reply xonsh/xontribs_meta.py:58]: I did some testings and amalgamate doesn't do much improvement on ssd. I think, unless there is some comment on removing amalgamate altogether. The imports can be placed on the top, they don't have much import cost.

**Review by gforsyth — APPROVED** [review]
looks good!  thanks for putting this in!

**Pull request merged**